### PR TITLE
Copter: Refactoring the code

### PR DESCRIPTION
**Pull request by muramura**
Stop making more decisions in order to have one layer of code.
It is true that the more layers, the harder it is to understand.
However, with this code, I think it would be better to make it a layer.

### CODE REVIEW

**Review by amilcarlucas — APPROVED** [review]

File: ArduCopter/GCS_Copter.cpp [hunk 1]
```diff
@@ -116,22 +116,24 @@ void GCS_Copter::update_vehicle_sensor_status_flags(void)
 
 #if OPTFLOW == ENABLED
     const OpticalFlow *optflow = AP::opticalflow();
-    if (optflow && optflow->enabled()) {
-        control_sensors_present |= MAV_SYS_STATUS_SENSOR_OPTICAL_FLOW;
-        control_sensors_enabled |= MAV_SYS_STATUS_SENSOR_OPTICAL_FLOW;
-    }
-    if (optflow && optflow->healthy()) {
-        control_sensors_health |= MAV_SYS_STATUS_SENSOR_OPTICAL_FLOW;
```
Comment (peterbarker) [inline ArduCopter/GCS_Copter.cpp:124]: This exact same code is present in Plane *and* Sub.  After this change we'll have two different versions of the code.

I do prefer your new formulation here.

.... but I also think we should probably put OpticalFlow on AP_Vehicle, especially now we can play with scheduler tables priority a bit better.

**Review by peterbarker — CHANGES_REQUESTED** [review]
Commit comment needs fixing:

`Copter: remove redundant checks when calculating SYS_STATUS flags`

**Comment by peterbarker** [comment]
... sorry, I should say you don't need to move OpticalFlow up as part of this PR!

**Comment by peterbarker** [comment]
This code has been moved up into `GCS_MAVLink/GCS.cpp` and is shared across all vehicles.  It uses a callback to determine the failsafe status, so I think this is all good now.

Thanks anyway!

**Pull request closed without merging**